transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java

```diff
@@ -311,19 +317,24 @@ private void invokeChannelRead(Object msg) {
         }
     }
 
+    private void invokeNextChannelRead(AbstractChannelHandlerContext next, Object msg) {
+        nextChannelReadInvoked = true;
+        next.invokeChannelRead(msg);
+    }
+
     @Override
     public ChannelHandlerContext fireChannelReadComplete() {
         final AbstractChannelHandlerContext next = findContextInbound();
         EventExecutor executor = next.executor();
         if (executor.inEventLoop()) {
-            next.invokeChannelReadComplete();
+            invokeNextChannelReadComplete(next);
         } else {
             Runnable task = next.invokeChannelReadCompleteTask;
             if (task == null) {
                 next.invokeChannelReadCompleteTask = task = new Runnable() {
                     @Override
                     public void run() {
-                        next.invokeChannelReadComplete();
+                        invokeNextChannelReadComplete(next);
```

Comment:
Hmm... How isvthis different to what we had before. If I understood you correct the issue existed also before



> Am 07.02.2015 um 07:03 schrieb Trustin Lee <notifications@github.com>:
> 
> In transport/src/main/java/io/netty/channel/AbstractChannelHandlerContext.java:
> 
> >          } else {
> >              Runnable task = next.invokeChannelReadCompleteTask;
> >              if (task == null) {
> >                  next.invokeChannelReadCompleteTask = task = new Runnable() {
> >                      @Override
> >                      public void run() {
> > -                        next.invokeChannelReadComplete();
> > +                        invokeNextChannelReadComplete(next);
> And if pipeline manipulation occurs after this task is cached, it will break things.
> 
> —
> Reply to this email directly or view it on GitHub.
>